databases/s3/database.go

```diff
@@ -0,0 +1,275 @@
+	}
+	return stagingID, nil
+}
+
+func (db *Database) StagingStatus(id uuid.UUID) (databases.StagingStatus, error) {
+	// check if staging request exists
+	_, ok := db.StagingRequests[id]
+	if !ok {
+		return databases.StagingStatusUnknown, fmt.Errorf("staging request %s not found", id.String())
+	}
+	// since no actual staging is done, just return completed status
+	return databases.StagingStatusSucceeded, nil
+}
+
+func (db *Database) Finalize(orcid string, id uuid.UUID) error {
+	return nil
+}
+
+func (db *Database) LocalUser(orcid string) (string, error) {
+	return "", fmt.Errorf("local user lookup not implemented for S3 database")
+}
+
+func (db *Database) Save() (databases.DatabaseSaveState, error) {
+	var buffer bytes.Buffer
+	enc := gob.NewEncoder(&buffer)
+	err := enc.Encode(db.StagingRequests)
+	if err != nil {
+		return databases.DatabaseSaveState{}, fmt.Errorf("error encoding S3 database state: %v", err)
+	}
+	return databases.DatabaseSaveState{
+		Name: "s3",
+		Data: buffer.Bytes(),
+	}, nil
+}
+
+func (db *Database) Load(state databases.DatabaseSaveState) error {
+	enc := gob.NewDecoder(bytes.NewReader(state.Data))
+	return enc.Decode(&db.StagingRequests)
+}
+
+//////////////
+// Internals
+//////////////
+
+// returns whether a file exists in the S3 bucket
+func (db *Database) fileExists(key string) (bool, error) {
+	_, err := db.Client.HeadObject(context.TODO(), &awsS3.HeadObjectInput{
+		Bucket: aws.String(db.Bucket),
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), "NotFound") || strings.Contains(err.Error(), "404") || strings.Contains(err.Error(), "NoSuchKey") || strings.Contains(err.Error(), "Not Found") {
+			return false, nil
+		} else {
+			return false, fmt.Errorf("error checking existence of S3 object %s: %v", key, err)
+		}
+	}
+	return true, nil
+}
+
+// returns all files that have the given prefix in the S3 bucket
+func (db *Database) listFilesWithPrefix(prefix string) ([]string, error) {
+	var files []string
+	contents, err := db.Client.ListObjectsV2(context.TODO(), &awsS3.ListObjectsV2Input{
+		Bucket: aws.String(db.Bucket),
+		Prefix: aws.String(prefix),
+	})
+	if err != nil {
+		return files, fmt.Errorf("error listing files with prefix %s in bucket %s: %v", prefix, db.Bucket, err)
+	}
+	for _, obj := range contents.Contents {
+		files = append(files, aws.ToString(obj.Key))
+	}
+	return files, nil
+}
+
+// returns a frictionless file descriptor for the given S3 object key
+func (db *Database) s3ObjectToDescriptor(key string) (map[string]any, error) {
+	// get object head to retrieve metadata
+	headOutput, err := db.Client.HeadObject(context.TODO(), &awsS3.HeadObjectInput{
+		Bucket: aws.String(db.Bucket),
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving metadata for S3 object %s: %v", key, err)
+	}
+
+	parts := strings.Split(key, "/")
+	descriptor := map[string]any{
+		"name":      parts[len(parts)-1],
```

Comment:
[`filepath.Base`](https://pkg.go.dev/path/filepath#Base) is another way to get this (just FYI)